plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java

```diff
@@ -473,7 +473,9 @@ private void testMergeWithSpecifiedRowkeys(String rowkeyDefinition)
         String targetTable = "merge_multiple_rowkeys_specified_" + randomNameSuffix();
         // check the upper case table name also works
         targetTable = targetTable.toUpperCase(ENGLISH);
-        assertUpdate(createTableForWrites(format("CREATE TABLE %s (customer VARCHAR, purchases INT, zipcode INT, spouse VARCHAR, address VARCHAR, customer_copy VARCHAR) WITH (rowkeys = '%s')", targetTable, rowkeyDefinition)));
+        // TODO: resolve performing merge operations with non-default rowkeys with the new merge paradigm.
```

Comment:
The only failure Iv'e encountered with the new paradigm, there is an issue when changing default rowkeys in merge operations:
```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 50.51 s <<< FAILURE! -- in io.trino.plugin.phoenix5.TestPhoenixConnectorTest
[ERROR] io.trino.plugin.phoenix5.TestPhoenixConnectorTest.testMergeWithSpecifiedRowkeys -- Time elapsed: 3.159 s <<< ERROR!
io.trino.testing.QueryFailedException: Failed to insert data: ERROR 1106 (XCL06): Exception while executing batch.
	at io.trino.testing.AbstractTestingTrinoClient.execute(AbstractTestingTrinoClient.java:133)
	at io.trino.testing.DistributedQueryRunner.executeWithQueryId(DistributedQueryRunner.java:519)
	at io.trino.testing.QueryAssertions.assertDistributedUpdate(QueryAssertions.java:107)
	at io.trino.testing.QueryAssertions.assertUpdate(QueryAssertions.java:63)
	at io.trino.testing.AbstractTestQueryFramework.assertUpdate(AbstractTestQueryFramework.java:419)
	at io.trino.testing.AbstractTestQueryFramework.assertUpdate(AbstractTestQueryFramework.java:414)
	at io.trino.plugin.phoenix5.TestPhoenixConnectorTest.testMergeWithSpecifiedRowkeys(TestPhoenixConnectorTest.java:491)
	at io.trino.plugin.phoenix5.TestPhoenixConnectorTest.testMergeWithSpecifiedRowkeys(TestPhoenixConnectorTest.java:463)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:728)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:218)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:214)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:139)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:69)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService$ExclusiveTask.compute(ForkJoinPoolHierarchicalTestExecutorService.java:202)
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService.invokeAll(ForkJoinPoolHierarchicalTestExecutorService.java:146)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService$ExclusiveTask.compute(ForkJoinPoolHierarchicalTestExecutorService.java:202)
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService.invokeAll(ForkJoinPoolHierarchicalTestExecutorService.java:146)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService$ExclusiveTask.compute(ForkJoinPoolHierarchicalTestExecutorService.java:202)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
	Suppressed: java.lang.Exception: SQL: MERGE INTO MERGE_MULTIPLE_ROWKEYS_SPECIFIED_TFAH7HJ9I1 t USING (VALUES ('joe_16', 3000, 83000, 'jill_16', '16 Eop Ct', 'joe_16'), ('joe_17', 3000, 83000, 'jill_17', '17 Eop Ct', 'joe_17'), ('joe_18', 3000, 83000, 'jill_18', '18 Eop Ct', 'joe_18'), ('joe_19', 3000, 83000, 'jill_19', '19 Eop Ct', 'joe_19'), ('joe_20', 3000, 83000, 'jill_20', '20 Eop Ct', 'joe_20'), ('joe_21', 3000, 83000, 'jill_21', '21 Eop Ct', 'joe_21'), ('joe_22', 3000, 83000, 'jill_22', '22 Eop Ct', 'joe_22'), ('joe_23', 3000, 83000, 'jill_23', '23 Eop Ct', 'joe_23'), ('joe_24', 3000, 83000, 'jill_24', '24 Eop Ct', 'joe_24'), ('joe_25', 3000, 83000, 'jill_25', '25 Eop Ct', 'joe_25'), ('joe_26', 3000, 83000, 'jill_26', '26 Eop Ct', 'joe_26'), ('joe_27', 3000, 83000, 'jill_27', '27 Eop Ct', 'joe_27'), ('joe_28', 3000, 83000, 'jill_28', '28 Eop Ct', 'joe_28'), ('joe_29', 3000, 83000, 'jill_29', '29 Eop Ct', 'joe_29'), ('joe_30', 3000, 83000, 'jill_30', '30 Eop Ct', 'joe_30'), ('joe_31', 3000, 83000, 'jill_31', '31 Eop Ct', 'joe_31')) AS s(customer, purchases, zipcode, spouse, address, customer_copy)    ON t.customer = s.customer    WHEN MATCHED AND t.purchases = 1234 THEN DELETE    WHEN MATCHED THEN UPDATE SET purchases = s.purchases, zipcode = s.zipcode, spouse = s.spouse, address = s.address
		at io.trino.testing.DistributedQueryRunner.executeWithQueryId(DistributedQueryRunner.java:523)
		... 64 more
Caused by: io.trino.spi.TrinoException: Failed to insert data: ERROR 1106 (XCL06): Exception while executing batch.
	at io.trino.plugin.jdbc.JdbcPageSink.finish(JdbcPageSink.java:225)
	at io.trino.plugin.phoenix5.PhoenixMergeSink.finish(PhoenixMergeSink.java:224)
	at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMergeSink.finish(ClassLoaderSafeConnectorMergeSink.java:52)
	at io.trino.operator.MergeWriterOperator.finish(MergeWriterOperator.java:196)
	at io.trino.operator.Driver.processInternal(Driver.java:413)
	at io.trino.operator.Driver.lambda$process$8(Driver.java:298)
	at io.trino.operator.Driver.tryWithLock(Driver.java:694)
	at io.trino.operator.Driver.process(Driver.java:290)
	at io.trino.operator.Driver.processForDuration(Driver.java:261)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:887)
	at io.trino.execution.executor.timesharing.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:187)
	at io.trino.execution.executor.timesharing.TimeSharingTaskExecutor$TaskRunner.run(TimeSharingTaskExecutor.java:565)
	at io.trino.$gen.Trino_testversion____20231204_154549_75.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:842)
Caused by: org.apache.phoenix.exception.BatchUpdateExecution: ERROR 1106 (XCL06): Exception while executing batch.
	at org.apache.phoenix.jdbc.PhoenixStatement.executeBatch(PhoenixStatement.java:1842)
	at io.opentelemetry.instrumentation.jdbc.internal.OpenTelemetryStatement.wrapCall(OpenTelemetryStatement.java:294)
	at io.opentelemetry.instrumentation.jdbc.internal.OpenTelemetryStatement.executeBatch(OpenTelemetryStatement.java:106)
	at io.trino.plugin.jdbc.JdbcPageSink.finish(JdbcPageSink.java:209)
	... 15 more
Caused by: java.sql.SQLException: ERROR 218 (23018): Constraint violation. TPCH.MERGE_MULTIPLE_ROWKEYS_SPECIFIED_TFAH7HJ9I1.CUSTOMER may not be null
	at org.apache.phoenix.exception.SQLExceptionCode$Factory$1.newException(SQLExceptionCode.java:607)
	at org.apache.phoenix.exception.SQLExceptionInfo.buildException(SQLExceptionInfo.java:217)
	at org.apache.phoenix.schema.ConstraintViolationException.<init>(ConstraintViolationException.java:40)
	at org.apache.phoenix.compile.UpsertCompiler.compile(UpsertCompiler.java:567)
	at org.apache.phoenix.jdbc.PhoenixStatement$ExecutableUpsertStatement.compilePlan(PhoenixStatement.java:847)
	at org.apache.phoenix.jdbc.PhoenixStatement$ExecutableUpsertStatement.compilePlan(PhoenixStatement.java:833)
	at org.apache.phoenix.jdbc.PhoenixStatement$2.call(PhoenixStatement.java:435)
	at org.apache.phoenix.jdbc.PhoenixStatement$2.call(PhoenixStatement.java:425)
	at org.apache.phoenix.call.CallRunner.run(CallRunner.java:53)
	at org.apache.phoenix.jdbc.PhoenixStatement.executeMutation(PhoenixStatement.java:424)
	at org.apache.phoenix.jdbc.PhoenixStatement.executeMutation(PhoenixStatement.java:412)
	at org.apache.phoenix.jdbc.PhoenixPreparedStatement.execute(PhoenixPreparedStatement.java:168)
	at org.apache.phoenix.jdbc.PhoenixStatement.executeBatch(PhoenixStatement.java:1834)
	... 18 more
```
Can someone from Phoenix assist with debugging this issue?

Reply:
I wonder if @chenjian2664 might be able to help here?  I see he made significant changes to the Phoenix connector in 2023.

Reply:
Ping @djsstarburst @chenjian2664 :)